File: backend/src/services/fileService.ts
import * as fs from 'fs';
import * as path from 'path';
import { DocNode, DiagnoseRecord, TimelineData } from '../types';

const BASE_DIR = path.join(__dirname, '../../..');

// 获取raw目录的根路径
export const getRawPath = (): string => {
  return path.join(BASE_DIR, 'raw');
};

// 获取财报目录根路径
export const getFinanceReportsPath = (): string => {
  return path.join(BASE_DIR, 'finance_reports');
};

// 获取研报目录根路径
export const getResearchReportsPath = (): string => {
  return path.join(BASE_DIR, 'research_reports');
};

// 获取各输出目录路径
export const getNewPath = (): string => path.join(BASE_DIR, 'new');
export const getReportPath = (): string => path.join(BASE_DIR, 'report');
export const getTimelinePath = (): string => path.join(BASE_DIR, 'timeline');

// 读取文档内容
export const readDocument = (relativePath: string): string => {
  const fullPath = path.join(getRawPath(), relativePath);
  return fs.readFileSync(fullPath, 'utf-8');
};

// 读取财报文档内容（finance_reports 下相对路径）
export const readFinanceDocument = (relativePath: string): string => {
  const fullPath = path.join(getFinanceReportsPath(), relativePath);
  return fs.readFileSync(fullPath, 'utf-8');
};
// 按文件后缀解析文档内容，避免二进制文件按 utf-8 读取导致乱码
const parseDocumentByExt = async (fullPath: string): Promise<string> => {
  const ext = path.extname(fullPath).toLowerCase();
  const textExts = new Set(['.md', '.markdown', '.txt', '.json', '.html', '.htm', '.csv', '.yaml', '.yml']);

  if (textExts.has(ext)) {
    return fs.readFileSync(fullPath, 'utf-8');
  }

  if (ext === '.pdf') {
    const pdfModule = await import('pdf-parse');
    const pdfParse = (pdfModule as any).default || (pdfModule as any);
    const buffer = fs.readFileSync(fullPath);
    const parsed = await pdfParse(buffer);
    return parsed?.text || '';
  }

  // 未知后缀默认按文本读取；若失败则给出可读提示，避免直接显示乱码
  try {
    return fs.readFileSync(fullPath, 'utf-8');
  } catch {
    return `暂不支持预览该文件类型：${ext || 'unknown'}`;
  }
};

export const readDocumentParsed = async (relativePath: string): Promise<string> => {
  const fullPath = path.join(getRawPath(), relativePath);
  return parseDocumentByExt(fullPath);
};

export const readFinanceDocumentParsed = async (relativePath: string): Promise<string> => {
  const fullPath = path.join(getFinanceReportsPath(), relativePath);
  return parseDocumentByExt(fullPath);
};

export const readResearchDocumentParsed = async (relativePath: string): Promise<string> => {
  const fullPath = path.join(getResearchReportsPath(), relativePath);
  return parseDocumentByExt(fullPath);
};

// 构建文档目录树
export const buildDocTree = (relativePath: string = ''): DocNode[] => {
  const rawRoot = getRawPath();
  const fullPath = path.join(rawRoot, relativePath);

  if (!fs.existsSync(rawRoot)) {
    console.warn('[buildDocTree] raw 目录不存在:', rawRoot);
    return [];
  }

  let items: string[];
  try {
    items = fs.readdirSync(fullPath);
  } catch (e) {
    console.error('[buildDocTree] 无法读取目录:', fullPath, e);
    return [];
  }

  const sortable = items
    .filter((item) => !item.startsWith('.'))
    .map((item) => {
      const itemPath = path.join(fullPath, item);
      try {
        const isDirectory = fs.statSync(itemPath).isDirectory();
        return { item, itemPath, isDirectory };
      } catch (e) {
        console.warn('[buildDocTree] 跳过无法访问的项:', itemPath, e);
        return null;
      }
    })
    .filter((x): x is NonNullable<typeof x> => x !== null)
    .sort((a, b) => {
      if (a.isDirectory && !b.isDirectory) return -1;
      if (!a.isDirectory && b.isDirectory) return 1;
      return a.item.localeCompare(b.item);
    });

  const nodes: DocNode[] = [];
  for (const { item, itemPath, isDirectory } of sortable) {
    const itemRelativePath = path.join(relativePath, item);

    const node: DocNode = {
      name: item,
      path: itemPath,
      relativePath: itemRelativePath,
      type: isDirectory ? 'directory' : 'file',
    };

    try {
      if (isDirectory) {
        node.children = buildDocTree(itemRelativePath);
      } else {
        const history = getDiagnoseHistory(itemRelativePath);
        node.diagnoseStatus = history.length > 0 ? 'has-history' : 'none';
        if (history.length > 0) {
          node.lastDiagnoseTime = history[0].timestamp;
        }
      }
    } catch (e) {
      console.error('[buildDocTree] 处理节点失败:', itemRelativePath, e);
      if (!isDirectory) {
        node.diagnoseStatus = 'none';
      } else {
        node.children = [];
      }
    }

    nodes.push(node);
  }

  return nodes;
};

// 构建 finance_reports 目录树（不附带诊断历史）
export const buildFinanceDocTree = (relativePath: string = ''): DocNode[] => {
  const financeRoot = getFinanceReportsPath();
  const fullPath = path.join(financeRoot, relativePath);

  if (!fs.existsSync(financeRoot)) {
    return [];
  }

  let items: string[];
  try {
    items = fs.readdirSync(fullPath);
  } catch (e) {
    console.error('[buildFinanceDocTree] 无法读取目录:', fullPath, e);
    return [];
  }

  const sortable = items
    .filter((item) => !item.startsWith('.'))
    .map((item) => {
      const itemPath = path.join(fullPath, item);
      try {
        const isDirectory = fs.statSync(itemPath).isDirectory();
        return { item, itemPath, isDirectory };
      } catch (e) {
        console.warn('[buildFinanceDocTree] 跳过无法访问的项:', itemPath, e);
        return null;
      }
    })
    .filter((x): x is NonNullable<typeof x> => x !== null)
    .sort((a, b) => {
      if (a.isDirectory && !b.isDirectory) return -1;
      if (!a.isDirectory && b.isDirectory) return 1;
      return a.item.localeCompare(b.item);
    });

  const nodes: DocNode[] = [];
  for (const { item, itemPath, isDirectory } of sortable) {
    const itemRelativePath = path.join(relativePath, item);
    const node: DocNode = {
      name: item,
      path: itemPath,
      relativePath: itemRelativePath,
      type: isDirectory ? 'directory' : 'file',
    };
    if (isDirectory) {
      node.children = buildFinanceDocTree(itemRelativePath);
    }
    nodes.push(node);
  }

  return nodes;
};

// 构建 research_reports 目录树（不附带诊断历史）
export const buildResearchDocTree = (relativePath: string = ''): DocNode[] => {
  const researchRoot = getResearchReportsPath();
  const fullPath = path.join(researchRoot, relativePath);

  if (!fs.existsSync(researchRoot)) {
    return [];
  }

  let items: string[];
  try {
    items = fs.readdirSync(fullPath);
  } catch (e) {
    console.error('[buildResearchDocTree] 无法读取目录:', fullPath, e);
    return [];
  }

  const sortable = items
    .filter((item) => !item.startsWith('.'))
    .map((item) => {
      const itemPath = path.join(fullPath, item);
      try {
        const isDirectory = fs.statSync(itemPath).isDirectory();
        return { item, itemPath, isDirectory };
      } catch (e) {
        console.warn('[buildResearchDocTree] 跳过无法访问的项:', itemPath, e);
        return null;
      }
    })
    .filter((x): x is NonNullable<typeof x> => x !== null)
    .sort((a, b) => {
      if (a.isDirectory && !b.isDirectory) return -1;
      if (!a.isDirectory && b.isDirectory) return 1;
      return a.item.localeCompare(b.item);
    });

  const nodes: DocNode[] = [];
  for (const { item, itemPath, isDirectory } of sortable) {
    const itemRelativePath = path.join(relativePath, item);
    const node: DocNode = {
      name: item,
      path: itemPath,
      relativePath: itemRelativePath,
      type: isDirectory ? 'directory' : 'file',
    };
    if (isDirectory) {
      node.children = buildResearchDocTree(itemRelativePath);
    }
    nodes.push(node);
  }

  return nodes;
};

// 获取指定文档的诊断历史
export const getDiagnoseHistory = (documentRelativePath: string): DiagnoseRecord[] => {
  const timelineDir = getTimelinePath();

  // 构建可能的timeline文件名模式
  const docName = path.basename(documentRelativePath, '.md');
  const docDir = path.dirname(documentRelativePath);

  if (!fs.existsSync(timelineDir)) {
    return [];
  }

  const records: DiagnoseRecord[] = [];

  // 遍历timeline目录查找匹配的文件
  const traverseDir = (dir: string) => {
    if (!fs.existsSync(dir)) return;

    let items: string[];
    try {
      items = fs.readdirSync(dir);
    } catch (e) {
      console.warn('[getDiagnoseHistory] 无法读取目录:', dir, e);
      return;
    }

    for (const item of items) {
      const itemPath = path.join(dir, item);
      let stat: fs.Stats;
      try {
        stat = fs.statSync(itemPath);
      } catch (e) {
        console.warn('[getDiagnoseHistory] 跳过无法 stat 的项:', itemPath, e);
        continue;
      }

      if (stat.isDirectory()) {
        traverseDir(itemPath);
      } else if (item.endsWith('_timeline.json')) {
        // 检查是否匹配当前文档
        // 文件名格式: xxx_20260427_143052_timeline.json
        // 需要匹配 xxx_日期_时间
        const baseName = item.replace('_timeline.json', '');
        if (baseName.startsWith(docName + '_')) {
          try {
            const content = fs.readFileSync(itemPath, 'utf-8');
            const timeline: TimelineData = JSON.parse(content);
            const summary = timeline.summary;
            const outputs = timeline.outputs;
            if (!summary?.by_severity || !outputs) {
              console.warn('[getDiagnoseHistory] timeline 缺少 summary/outputs，跳过:', itemPath);
              continue;
            }

            const reportPath = outputs.report
              ? path.join(BASE_DIR, outputs.report)
              : '';
            const fixedDocPath = outputs.fixed_doc
              ? path.join(BASE_DIR, outputs.fixed_doc)
              : '';

            records.push({
              timestamp: (timeline.start_time || '').replace('T', ' ').split('.')[0],
              documentPath: documentRelativePath,
              documentName: path.basename(documentRelativePath),
              status: timeline.status,
              totalIssues: summary.total_issues ?? 0,
              highPriority: summary.by_severity.high ?? 0,
              mediumPriority: summary.by_severity.medium ?? 0,
              lowPriority: summary.by_severity.low ?? 0,
              reportPath,
              timelinePath: itemPath,
              fixedDocPath,
              durationSeconds: timeline.duration_seconds,
            });
          } catch (e) {
            console.error('Failed to parse timeline:', itemPath, e);
          }
        }
      }
    }
  };

  // 先在对应子目录下查找
  const targetDir = path.join(timelineDir, docDir);
  traverseDir(targetDir);

  // 如果没找到，遍历整个timeline目录
  if (records.length === 0) {
    traverseDir(timelineDir);
  }

  // 按时间倒序排列
  return records.sort((a, b) => b.timestamp.localeCompare(a.timestamp));
};

// 获取指定财报文档的分析历史（timeline/finance）
export const getFinanceHistory = (documentRelativePath: string): DiagnoseRecord[] => {
  const financeTimelineDir = path.join(getTimelinePath(), 'finance');
  const fileName = path.basename(documentRelativePath);
  const fileNameNoExt = path.basename(documentRelativePath, path.extname(documentRelativePath));
  const expectedDocPath = `finance_reports/${documentRelativePath}`.replace(/\\/g, '/');
  const records: DiagnoseRecord[] = [];

  const walkTimeline = (dir: string) => {
    if (!fs.existsSync(dir)) return;
    const items = fs.readdirSync(dir);
    for (const item of items) {
      const itemPath = path.join(dir, item);
      const stat = fs.statSync(itemPath);
      if (stat.isDirectory()) {
        walkTimeline(itemPath);
        continue;
      }
      if (!item.endsWith('_timeline.json')) {
        continue;
      }

      try {
        const content = fs.readFileSync(itemPath, 'utf-8');
        const timeline: TimelineData = JSON.parse(content);
        const timelineDocPath = (timeline.document?.path || '').replace(/\\/g, '/');
        const summary = timeline.summary;
        const outputs = timeline.outputs;
        if (!summary?.by_severity || !outputs) {
          continue;
        }

        const matchedByDocPath =
          timelineDocPath === expectedDocPath ||
          timelineDocPath.endsWith(`/${documentRelativePath.replace(/\\/g, '/')}`) ||
          timelineDocPath.endsWith(`/${fileName}`) ||
          timelineDocPath === documentRelativePath.replace(/\\/g, '/');
        const matchedByTimelineName = item.startsWith(`${fileNameNoExt}_`);
        if (!matchedByDocPath && !matchedByTimelineName) {
          continue;
        }

        const reportPath = outputs.report ? path.join(BASE_DIR, outputs.report) : '';
        const fixedDocPath = outputs.fixed_doc ? path.join(BASE_DIR, outputs.fixed_doc) : '';

        records.push({
          timestamp: (timeline.start_time || '').replace('T', ' ').split('.')[0],
          documentPath: documentRelativePath,
          documentName: fileName,
          status: timeline.status,
          totalIssues: summary.total_issues ?? 0,
          highPriority: summary.by_severity.high ?? 0,
          mediumPriority: summary.by_severity.medium ?? 0,
          lowPriority: summary.by_severity.low ?? 0,
          reportPath,
          timelinePath: itemPath,
          fixedDocPath,
          durationSeconds: timeline.duration_seconds,
        });
      } catch (e) {
        console.error('Failed to parse finance timeline:', itemPath, e);
      }
    }
  };

  const collectFromReports = (): DiagnoseRecord[] => {
    const reportRoot = path.join(getReportPath(), 'finance', 'cn-a-share', 'risk-opportunity');
    if (!fs.existsSync(reportRoot)) return [];

    const fallbackRecords: DiagnoseRecord[] = [];
    const walkReport = (dir: string) => {
      const items = fs.readdirSync(dir);
      for (const item of items) {
        const itemPath = path.join(dir, item);
        const stat = fs.statSync(itemPath);
        if (stat.isDirectory()) {
          walkReport(itemPath);
          continue;
        }
        if (!item.endsWith('_report.md')) continue;

        try {
          const content = fs.readFileSync(itemPath, 'utf-8');
          const normalizedContent = content.replace(/\\/g, '/');
          const targetA = `finance_reports/${documentRelativePath.replace(/\\/g, '/')}`;
          const targetB = `\`${targetA}\``;
          if (!normalizedContent.includes(targetA) && !normalizedContent.includes(targetB)) {
            continue;
          }

          const timestampMatch = item.match(/_(\d{8}_\d{6})_report\.md$/);
          const rawTs = timestampMatch?.[1] || '';
          const timestamp = rawTs
            ? `${rawTs.slice(0, 4)}-${rawTs.slice(4, 6)}-${rawTs.slice(6, 8)} ${rawTs.slice(9, 11)}:${rawTs.slice(11, 13)}:${rawTs.slice(13, 15)}`
            : '';

          fallbackRecords.push({
            timestamp,
            documentPath: documentRelativePath,
            documentName: fileName,
            status: 'completed',
            totalIssues: 0,
            highPriority: 0,
            mediumPriority: 0,
            lowPriority: 0,
            reportPath: itemPath,
            timelinePath: '',
            fixedDocPath: '',
          });
        } catch (e) {
          console.error('Failed to parse finance report for history fallback:', itemPath, e);
        }
      }
    };

    walkReport(reportRoot);
    return fallbackRecords;
  };

  if (fs.existsSync(financeTimelineDir)) {
    walkTimeline(financeTimelineDir);
  }

  if (records.length === 0) {
    records.push(...collectFromReports());
  }

  return records.sort((a, b) => (b.timestamp || '').localeCompare(a.timestamp || ''));
};

// 获取指定研报文档的分析历史（timeline/research）
export const getResearchHistory = (documentRelativePath: string): DiagnoseRecord[] => {
  const researchTimelineDir = path.join(getTimelinePath(), 'research');
  const fileName = path.basename(documentRelativePath);
  const fileNameNoExt = path.basename(documentRelativePath, path.extname(documentRelativePath));
  const expectedDocPath = `research_reports/${documentRelativePath}`.replace(/\\/g, '/');
  const records: DiagnoseRecord[] = [];

  const walkTimeline = (dir: string) => {
    if (!fs.existsSync(dir)) return;
    const items = fs.readdirSync(dir);
    for (const item of items) {
      const itemPath = path.join(dir, item);
      const stat = fs.statSync(itemPath);
      if (stat.isDirectory()) {
        walkTimeline(itemPath);
        continue;
      }
      if (!item.endsWith('_timeline.json')) {
        continue;
      }

      try {
        const content = fs.readFileSync(itemPath, 'utf-8');
        const timeline: TimelineData = JSON.parse(content);
        const timelineDocPath = (timeline.document?.path || '').replace(/\\/g, '/');
        const summary = timeline.summary;
        const outputs = timeline.outputs;
        if (!summary?.by_severity || !outputs) {
          continue;
        }

        const matchedByDocPath =
          timelineDocPath === expectedDocPath ||
          timelineDocPath.endsWith(`/${documentRelativePath.replace(/\\/g, '/')}`) ||
          timelineDocPath.endsWith(`/${fileName}`) ||
          timelineDocPath === documentRelativePath.replace(/\\/g, '/');
        const matchedByTimelineName = item.startsWith(`${fileNameNoExt}_`);
        if (!matchedByDocPath && !matchedByTimelineName) {
          continue;
        }

        const reportPath = outputs.report ? path.join(BASE_DIR, outputs.report) : '';
        const fixedDocPath = outputs.fixed_doc ? path.join(BASE_DIR, outputs.fixed_doc) : '';

        records.push({
          timestamp: (timeline.start_time || '').replace('T', ' ').split('.')[0],
          documentPath: documentRelativePath,
          documentName: fileName,
          status: timeline.status,
          totalIssues: summary.total_issues ?? 0,
          highPriority: summary.by_severity.high ?? 0,
          mediumPriority: summary.by_severity.medium ?? 0,
          lowPriority: summary.by_severity.low ?? 0,
          reportPath,
          timelinePath: itemPath,
          fixedDocPath,
          durationSeconds: timeline.duration_seconds,
        });
      } catch (e) {
        console.error('Failed to parse research timeline:', itemPath, e);
      }
    }
  };

  const collectFromReports = (): DiagnoseRecord[] => {
    const reportRoot = path.join(getReportPath(), 'research', 'cn-a-share');
    if (!fs.existsSync(reportRoot)) return [];

    const fallbackRecords: DiagnoseRecord[] = [];
    const walkReport = (dir: string) => {
      const items = fs.readdirSync(dir);
      for (const item of items) {
        const itemPath = path.join(dir, item);
        const stat = fs.statSync(itemPath);
        if (stat.isDirectory()) {
          walkReport(itemPath);
          continue;
        }
        if (!item.endsWith('_report.md')) continue;

        try {
          const content = fs.readFileSync(itemPath, 'utf-8');
          const normalizedContent = content.replace(/\\/g, '/');
          const targetA = `research_reports/${documentRelativePath.replace(/\\/g, '/')}`;
          const targetB = `\`${targetA}\``;
          if (!normalizedContent.includes(targetA) && !normalizedContent.includes(targetB)) {
            continue;
          }

          const timestampMatch = item.match(/_(\d{8}_\d{6})_report\.md$/);
          const rawTs = timestampMatch?.[1] || '';
          const timestamp = rawTs
            ? `${rawTs.slice(0, 4)}-${rawTs.slice(4, 6)}-${rawTs.slice(6, 8)} ${rawTs.slice(9, 11)}:${rawTs.slice(11, 13)}:${rawTs.slice(13, 15)}`
            : '';

          fallbackRecords.push({
            timestamp,
            documentPath: documentRelativePath,
            documentName: fileName,
            status: 'completed',
            totalIssues: 0,
            highPriority: 0,
            mediumPriority: 0,
            lowPriority: 0,
            reportPath: itemPath,
            timelinePath: '',
            fixedDocPath: '',
          });
        } catch (e) {
          console.error('Failed to parse research report for history fallback:', itemPath, e);
        }
      }
    };

    walkReport(reportRoot);
    return fallbackRecords;
  };

  if (fs.existsSync(researchTimelineDir)) {
    walkTimeline(researchTimelineDir);
  }

  if (records.length === 0) {
    records.push(...collectFromReports());
  }

  return records.sort((a, b) => (b.timestamp || '').localeCompare(a.timestamp || ''));
};

// 读取timeline文件
export const readTimeline = (timelinePath: string): TimelineData | null => {
  try {
    const content = fs.readFileSync(timelinePath, 'utf-8');
    return JSON.parse(content);
  } catch (e) {
    console.error('Failed to read timeline:', timelinePath, e);
    return null;
  }
};

// 读取报告文件
export const readReport = (reportPath: string): string => {
  try {
    // 如果是相对路径（不以/开头），则转换为绝对路径
    let fullPath = reportPath;
    if (!reportPath.startsWith('/')) {
      fullPath = path.join(BASE_DIR, reportPath);
    }

    if (!fs.existsSync(fullPath)) {
      return '报告文件不存在';
    }
    return fs.readFileSync(fullPath, 'utf-8');
  } catch (e) {
    return '报告文件不存在';
  }
};

// 读取修复后的文档
export const readFixedDoc = (fixedDocPath: string): string => {
  try {
    // 如果是相对路径（不以/开头），则转换为绝对路径
    let fullPath = fixedDocPath;
    if (!fixedDocPath.startsWith('/')) {
      fullPath = path.join(BASE_DIR, fixedDocPath);
    }

    if (!fs.existsSync(fullPath)) {
      return '修复后的文档不存在';
    }
    return fs.readFileSync(fullPath, 'utf-8');
  } catch (e) {
    return '修复后的文档不存在';
  }
};

// 确保目录存在
export const ensureDir = (dirPath: string): void => {
  if (!fs.existsSync(dirPath)) {
    fs.mkdirSync(dirPath, { recursive: true });
  }
};

// 保存修复后的文档
export const writeFixedDoc = (fixedDocPath: string, content: string): void => {
  let fullPath = fixedDocPath;
  if (!fixedDocPath.startsWith('/')) {
    fullPath = path.join(BASE_DIR, fixedDocPath);
  }
  const dir = path.dirname(fullPath);
  ensureDir(dir);
  fs.writeFileSync(fullPath, content, 'utf-8');
};

File: desktop/src-tauri/src/services/file_service.rs
use crate::models::*;
use std::fs;
use std::path::PathBuf;
use std::sync::Mutex;
use tauri::State;
use walkdir::WalkDir;

pub type WorkDirState = Mutex<Option<String>>;

/// 获取工作目录
pub fn get_work_dir(work_dir: &State<WorkDirState>) -> Option<PathBuf> {
    let dir = work_dir.lock().ok()?;
    dir.as_ref().map(|d| PathBuf::from(d))
}

/// 获取 raw 目录路径
fn get_raw_path(work_dir: &State<WorkDirState>) -> Option<PathBuf> {
    get_work_dir(work_dir).map(|d| d.join("raw"))
}

/// 获取 timeline 目录路径
fn get_timeline_path(work_dir: &State<WorkDirState>) -> Option<PathBuf> {
    get_work_dir(work_dir).map(|d| d.join("timeline"))
}

/// 安全地将路径转换为字符串
fn path_to_string(path: &PathBuf) -> String {
    path.to_str().unwrap_or("").to_string()
}

/// 安全地将 OsStr 转换为字符串
fn osstr_to_string(s: &std::ffi::OsStr) -> String {
    s.to_str().unwrap_or("").to_string()
}

/// 构建文档目录树
pub fn build_doc_tree(work_dir: &State<WorkDirState>) -> Result<Vec<DocNode>, String> {
    let raw_path = get_raw_path(work_dir).ok_or("工作目录未设置")?;
    
    if !raw_path.exists() {
        return Err(format!("raw 目录不存在: {:?}", raw_path));
    }
    
    Ok(build_doc_tree_recursive(&raw_path, &raw_path, work_dir))
}

fn build_doc_tree_recursive(
    current_path: &PathBuf,
    raw_path: &PathBuf,
    work_dir: &State<WorkDirState>,
) -> Vec<DocNode> {
    let entries = match fs::read_dir(current_path) {
        Ok(entries) => entries,
        Err(_) => return vec![],
    };

    let mut nodes: Vec<DocNode> = entries
        .filter_map(|entry| entry.ok())
        .filter(|entry| {
            let name = entry.file_name();
            name.to_str().map(|s| !s.starts_with('.')).unwrap_or(false)
        })
        .map(|entry| {
            let path = entry.path();
            let relative_path = match path.strip_prefix(raw_path) {
                Ok(p) => p.to_str().unwrap_or("").to_string(),
                Err(_) => "".to_string(),
            };
            let is_dir = path.is_dir();
            
            let name = osstr_to_string(&entry.file_name());
            
            let mut node = DocNode {
                name: name.clone(),
                path: path_to_string(&path),
                relative_path: relative_path.clone(),
                node_type: if is_dir { "directory".to_string() } else { "file".to_string() },
                children: None,
                diagnose_status: None,
                last_diagnose_time: None,
                history: None,
            };

            if is_dir {
                node.children = Some(build_doc_tree_recursive(&path, raw_path, work_dir));
            } else if path.extension().map(|e| e == "md").unwrap_or(false) {
                // 检查诊断历史
                let history = get_diagnose_history(&relative_path, work_dir);
                if !history.is_empty() {
                    node.diagnose_status = Some("has-history".to_string());
                    node.last_diagnose_time = Some(history[0].timestamp.clone());
                } else {
                    node.diagnose_status = Some("none".to_string());
                }
            }

            node
        })
        .collect();

    // 排序：目录在前，然后按名称排序
    nodes.sort_by(|a, b| {
        let a_is_dir = a.node_type == "directory";
        let b_is_dir = b.node_type == "directory";
        if a_is_dir && !b_is_dir {
            std::cmp::Ordering::Less
        } else if !a_is_dir && b_is_dir {
            std::cmp::Ordering::Greater
        } else {
            a.name.cmp(&b.name)
        }
    });

    nodes
}

/// 读取文档内容
pub fn read_document(relative_path: &str, work_dir: &State<WorkDirState>) -> Result<String, String> {
    let raw_path = get_raw_path(work_dir).ok_or("工作目录未设置")?;
    let full_path = raw_path.join(relative_path);
    
    fs::read_to_string(&full_path).map_err(|e| format!("读取文档失败: {}", e))
}

/// 获取诊断历史
pub fn get_diagnose_history(
    document_relative_path: &str,
    work_dir: &State<WorkDirState>,
) -> Vec<DiagnoseRecord> {
    let timeline_path = match get_timeline_path(work_dir) {
        Some(p) if p.exists() => p,
        _ => return vec![],
    };

    let doc_name = PathBuf::from(document_relative_path)
        .file_stem()
        .map(|s| s.to_str().unwrap_or(""))
        .unwrap_or("")
        .to_string();

    let mut records: Vec<DiagnoseRecord> = vec![];

    // 查找匹配的 timeline 文件
    for entry in WalkDir::new(&timeline_path)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| {
            e.file_name()
                .to_str()
                .map(|s| s.ends_with("_timeline.json"))
                .unwrap_or(false)
        })
    {
        let file_name = osstr_to_string(entry.file_name());
        let base_name = file_name.replace("_timeline.json", "");
        
        // 检查是否匹配当前文档
        let pattern = format!("{}_", doc_name);
        if base_name.starts_with(&pattern) {
            if let Ok(content) = fs::read_to_string(entry.path()) {
                if let Ok(timeline) = serde_json::from_str::<TimelineData>(&content) {
                    let base_dir = get_work_dir(work_dir);
                    let report_path = base_dir.as_ref()
                        .map(|d| {
                            d.join(&timeline.outputs.report)
                                .to_str()
                                .unwrap_or("")
                                .to_string()
                        })
                        .unwrap_or_default();
                    let fixed_doc_path = base_dir.as_ref()
                        .map(|d| {
                            d.join(&timeline.outputs.fixed_doc)
                                .to_str()
                                .unwrap_or("")
                                .to_string()
                        })
                        .unwrap_or_default();

                    records.push(DiagnoseRecord {
                        timestamp: timeline.start_time.split('.').next().unwrap_or(&timeline.start_time).replace('T', " "),
                        document_path: document_relative_path.to_string(),
                        document_name: PathBuf::from(document_relative_path)
                            .file_name()
                            .map(|s| s.to_str().unwrap_or(""))
                            .unwrap_or("")
                            .to_string(),
                        status: timeline.status,
                        total_issues: timeline.summary.total_issues,
                        high_priority: timeline.summary.by_severity.high,
                        medium_priority: timeline.summary.by_severity.medium,
                        low_priority: timeline.summary.by_severity.low,
                        report_path,
                        timeline_path: entry.path().to_str().unwrap_or("").to_string(),
                        fixed_doc_path,
                        duration_seconds: Some(timeline.duration_seconds),
                    });
                }
            }
        }
    }

    // 按时间倒序排列
    records.sort_by(|a, b| b.timestamp.cmp(&a.timestamp));
    records
}

/// 读取 Timeline 文件
pub fn read_timeline(timeline_path: &str) -> Result<TimelineData, String> {
    let content = fs::read_to_string(timeline_path)
        .map_err(|e| format!("读取 Timeline 失败: {}", e))?;
    
    serde_json::from_str(&content).map_err(|e| format!("解析 Timeline 失败: {}", e))
}

/// 读取报告文件
pub fn read_report(report_path: &str, work_dir: &State<WorkDirState>) -> Result<String, String> {
    let base_dir = get_work_dir(work_dir).ok_or("工作目录未设置")?;
    
    let full_path = if report_path.starts_with('/') {
        PathBuf::from(report_path)
    } else {
        base_dir.join(report_path)
    };

    if !full_path.exists() {
        return Ok("报告文件不存在".to_string());
    }

    fs::read_to_string(&full_path).map_err(|e| format!("读取报告失败: {}", e))
}

/// 读取修复后的文档
pub fn read_fixed_doc(fixed_doc_path: &str, work_dir: &State<WorkDirState>) -> Result<String, String> {
    let base_dir = get_work_dir(work_dir).ok_or("工作目录未设置")?;
    
    let full_path = if fixed_doc_path.starts_with('/') {
        PathBuf::from(fixed_doc_path)
    } else {
        base_dir.join(fixed_doc_path)
    };

    if !full_path.exists() {
        return Ok("修复后的文档不存在".to_string());
    }

    fs::read_to_string(&full_path).map_err(|e| format!("读取修复后文档失败: {}", e))
}

/// 保存修复后的文档
pub fn write_fixed_doc(fixed_doc_path: &str, content: &str, work_dir: &State<WorkDirState>) -> Result<String, String> {
    let base_dir = get_work_dir(work_dir).ok_or("工作目录未设置")?;
    
    let full_path = if fixed_doc_path.starts_with('/') {
        PathBuf::from(fixed_doc_path)
    } else {
        base_dir.join(fixed_doc_path)
    };

    // 确保目录存在
    if let Some(parent) = full_path.parent() {
        fs::create_dir_all(parent).map_err(|e| format!("创建目录失败: {}", e))?;
    }

    fs::write(&full_path, content).map_err(|e| format!("保存文件失败: {}", e))?;
    
    Ok("保存成功".to_string())
}